pkg/hub/controllers/cluster/deregister.go

```diff
@@ -0,0 +1,330 @@
+		return err
+	}
+	// Create a cluster role.
+	if err := r.MeshClient.Create(ctx, constructClusterRole(operatorClusterRole), &client.CreateOptions{}); err != nil {
+		if apierrors.IsAlreadyExists(err) {
+			log.Info("cluster role already exists", "clusterrole", clusterRoleName)
+		} else {
+			log.Error(err, "unable to create cluster role")
+			return err
+		}
+	}
+	// Bind the service account with the cluster role.
+	if err := r.MeshClient.Create(ctx, constructClusterRoleBinding(), &client.CreateOptions{}); err != nil {
+		if apierrors.IsAlreadyExists(err) {
+			log.Info("cluster rolebinding already exists", "clusterrolebinding", clusterRoleBindingName)
+		} else {
+			log.Error(err, "unable to create cluster rolebinding")
+			return err
+		}
+	}
+	// get configmap data
+	data, err := getConfigmapData()
+	if err != nil {
+		log.Error(err, "unable to fetch configmap data")
+		return err
+	}
+	// Set up a configuration map which contains the script for the job.
+	if err := r.MeshClient.Create(ctx, constructConfigMap(data), &client.CreateOptions{}); err != nil {
+		if apierrors.IsAlreadyExists(err) {
+			log.Info("cluster configmap already exists", "configmap", clusterDeregisterConfigMap)
+		} else {
+			log.Error(err, "unable to create cluster configmap")
+			return err
+		}
+	}
+
+	// get the job
+	jobRef := types.NamespacedName{
+		Name:      deregisterJobName,
+		Namespace: ControlPlaneNamespace,
+	}
+	deregisterJob := constructJobForClusterDeregister()
+	if err := r.MeshClient.Get(ctx, jobRef, deregisterJob); err == nil {
+		// Job is already present we need to delete it
+		err = r.MeshClient.Delete(ctx, deregisterJob)
+		if err != nil {
+			log.Error(err, "error while deleting job object", "job", jobRef.Name)
+			return err
+		}
+	}
+
+	// Generate a job for deregistration with the configuration map mounted as a volume.
+	if err := r.MeshClient.Create(ctx, deregisterJob, &client.CreateOptions{}); err != nil {
+		if apierrors.IsAlreadyExists(err) {
+			log.Info("cluster deregister job already exists", "job", clusterDeregisterConfigMap)
+		} else {
+			log.Error(err, "unable to create job for cluster deregister")
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *Reconciler) updateRegistrationStatusAndDeregisterInProgress(ctx context.Context, cluster *hubv1alpha1.Cluster, status hubv1alpha1.RegistrationStatus, isDeregisterInProgress bool) error {
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		err := r.Get(ctx, types.NamespacedName{
+			Name:      cluster.Name,
+			Namespace: cluster.Namespace,
+		}, cluster)
+		if err != nil {
+			return err
+		}
+		cluster.Status.RegistrationStatus = status
+		cluster.Status.IsDeregisterInProgress = isDeregisterInProgress
+		return r.Status().Update(ctx, cluster)
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *Reconciler) updateClusterDeregisterStatus(status hubv1alpha1.RegistrationStatus, ctx context.Context, cluster *hubv1alpha1.Cluster) error {
```

Comment:
Can we club both updateRegistrationStatusAndDeregisterInProgress and updateClusterDeregisterStatus methods
As updateRegistrationStatusAndDeregisterInProgress is sub status of the updateClusterDeregisterStatus

Reply:
We tried clubbing the functions together tests are failing because of the same function name. so currently we have segregated functions.

Reply:
I don't think this is right approach I think we can write UTs for a method of many lines